lib/auth/auth_with_roles.go

```diff
@@ -1257,14 +1278,45 @@ func (a *ServerWithRoles) generateUserCerts(ctx context.Context, req proto.UserC
 		roles = utils.Deduplicate(roles)
 	}
 
-	// Extract the user and role set for whom the certificate will be generated.
-	user, err := a.GetUser(req.Username, false)
+	parsedRoles, err := services.FetchRoleList(roles, a.authServer, traits)
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
-	checker, err := services.FetchRoles(roles, a.authServer, traits)
-	if err != nil {
-		return nil, trace.Wrap(err)
+	// add implicit roles to the set and build a checker
+	checker := services.NewRoleSet(parsedRoles...)
+
+	switch {
+	case a.hasBuiltinRole(string(teleport.RoleAdmin)):
+		// builtin admins can impersonate anyone
+		// this is required for local tctl commands to work
+	case req.Username == a.context.User.GetName():
+		// users can impersonate themselves
+	default:
+		// check if this user is allowed to impersonate other users
+		err = a.context.Checker.CheckImpersonate(a.context.User, user, parsedRoles)
+		// adjust session TTL based on the impersonated role set limit
+		ttl := req.Expires.Sub(a.authServer.GetClock().Now())
+		ttl = checker.AdjustSessionTTL(ttl)
+		req.Expires = a.authServer.GetClock().Now().Add(ttl)
+		if err != nil {
+			log.Warning(err)
+			err := trace.AccessDenied("user %q has requested to generate certs for %q.", a.context.User.GetName(), roles)
```

Comment:
How will admins using Cloud debug this?

I don't have a good answer to that question at the moment either, but @alex-kovoy has mentioned that he has to fetch process logs to help customers debug SAML connectors. I imagine a similar situation will happen here.